File: python/acronym/acronym.py


def abbreviate(phrase):
    return ''.join(w[0].upper() for w in _words(phrase))


def _words(phrase):
    words = []
    i = 0
    while i < len(phrase):
        if not phrase[i].isalpha():
            i += 1
        else:
            j = i
            while j < len(phrase) and phrase[j].isupper():
                j += 1
            while j < len(phrase) and phrase[j].islower():
                j += 1
            words.append(phrase[i:j])
            i = j

    return words

File: go/acronym/acronym.go
package acronym

import "unicode"

const testVersion = 2

func Abbreviate(s string) string {
	var words = _words(s)
	var result []rune
	for _, w := range words {
		result = append(result, unicode.ToUpper(w[0]))
	}
	return string(result)
}

func _words(s string) [][]rune {
	var phrase = []rune(s)
	var result [][]rune
	var i = 0
	for i < len(phrase) {
		if !unicode.IsLetter(phrase[i]) {
			i += 1
		} else {
			var j = i
			for j < len(phrase) && unicode.IsUpper(phrase[j]) {
				j += 1
			}
			for j < len(phrase) && unicode.IsLower(phrase[j]) {
				j += 1
			}
			result = append(result, phrase[i:j])
			i = j
		}
	}
	return result
}
